compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/FunctionValidator.java

```diff
@@ -283,7 +343,12 @@ private void validateReturnTypeUnion(List<TypeSymbol> returnTypeMembers, Locatio
                         if (!isRecordType(returnType)) {
                             updateContext(context, CompilationError.INVALID_RETURN_TYPE, location);
                         } else {
-                            recordTypes++;
+                            if (existInputObjectTypes.contains(returnType)) {
+                                updateContext(context, CompilationError.INVALID_RETURN_TYPE_INPUT_OBJECT, location);
+                            } else {
+                                existReturnTypes.add(returnType);
+                                recordTypes++;
+                            }
```

Comment:
This should be increased disregarding the outcome of the if condition, right?

```suggestion
                            }
                            recordTypes++;
```